File: App/Composition/ComposeTextViewController.m
//  ComposeTextViewController.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "ComposeTextViewController.h"
@import AssetsLibrary;
#import "AwfulFrameworkCategories.h"
#import "AwfulTextAttachment.h"
#import "ComposeTextView.h"
#import <ImgurAnonymousAPIClient/ImgurAnonymousAPIClient.h>
#import <MRProgress/MRProgressOverlayView.h>
#import "UploadImageAttachments.h"
#import "Awful-Swift.h"

@interface ComposeTextViewController ()

@property (readonly, strong, nonatomic) UIView *viewToOverlay;
@property (strong, nonatomic) CompositionMenuTree *menuTree;

@end

@implementation ComposeTextViewController
{
    UIBarButtonItem *_submitButtonItem;
    UIBarButtonItem *_cancelButtonItem;
    id _keyboardWillShowObserver;
    id _keyboardWillHideObserver;
    id _textDidChangeObserver;
    NSProgress *_imageUploadProgress;
    NSLayoutConstraint *_customViewWidthConstraint;
}

- (void)dealloc
{
    [self endObservingKeyboardNotifications];
    [self endObservingTextChangeNotification];
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    if ((self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil])) {
        self.navigationItem.leftBarButtonItem = self.cancelButtonItem;
        self.navigationItem.rightBarButtonItem = self.submitButtonItem;
    }
    return self;
}

- (UIBarButtonItem *)submitButtonItem
{
    if (_submitButtonItem) return _submitButtonItem;
    _submitButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Submit"
                                                         style:UIBarButtonItemStylePlain
                                                        target:self
                                                        action:@selector(didTapSubmit)];
    return _submitButtonItem;
}

- (UIBarButtonItem *)cancelButtonItem
{
    if (_cancelButtonItem) return _cancelButtonItem;
    _cancelButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Cancel"
                                                         style:UIBarButtonItemStylePlain
                                                        target:self
                                                        action:@selector(didTapCancel)];
    return _cancelButtonItem;
}

- (UITextView *)textView
{
    return (UITextView *)self.view;
}

- (void)loadView
{
    UITextView *textView = [ComposeTextView new];
    textView.restorationIdentifier = @"AwfulComposeTextView";
    textView.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
	textView.delegate = self;
    self.view = textView;
}

- (void)themeDidChange
{
    [super themeDidChange];
    self.textView.textColor = self.theme[@"listTextColor"];
    self.textView.keyboardAppearance = self.theme.keyboardAppearance;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.menuTree = [[CompositionMenuTree alloc] initWithTextView:self.textView];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self focusInitialFirstResponder];
    [self updateSubmitButtonItem];
    [self beginObservingKeyboardNotifications];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self beginObservingTextChangeNotification];
}

- (void)updateSubmitButtonItem
{
    self.submitButtonItem.enabled = self.canSubmitComposition;
}

- (BOOL)canSubmitComposition
{
    return self.textView.text.length > 0;
}

- (void)beginObservingTextChangeNotification
{
    _textDidChangeObserver = [[NSNotificationCenter defaultCenter] addObserverForName:UITextViewTextDidChangeNotification
                                                                               object:self.textView
                                                                                queue:[NSOperationQueue mainQueue]
                                                                           usingBlock:^(NSNotification *note)
    {
        [self updateSubmitButtonItem];
    }];
}

- (void)endObservingTextChangeNotification
{
    if (_textDidChangeObserver) {
        [[NSNotificationCenter defaultCenter] removeObserver:_textDidChangeObserver];
        _textDidChangeObserver = nil;
    }
}

- (void)shouldSubmitHandler:(void(^)(BOOL ok))handler
{
    handler(YES);
}

- (void)didTapSubmit
{
    [self disableEverythingButTheCancelButton];
    __weak __typeof__(self) weakSelf = self;
    [self shouldSubmitHandler:^(BOOL ok) {
        __typeof__(self) self = weakSelf;
        if (ok) {
            [self submit];
        } else {
            [_imageUploadProgress cancel];
            _imageUploadProgress = nil;
            [self enableEverything];
            [self focusInitialFirstResponder];
        }
    }];
}

- (void)submit
{
    MRProgressOverlayView *overlay = [MRProgressOverlayView showOverlayAddedTo:self.viewToOverlay
                                                                         title:self.submissionInProgressTitle
                                                                          mode:MRProgressOverlayViewModeIndeterminate
                                                                      animated:YES];
    overlay.tintColor = self.theme[@"tintColor"];

    __weak __typeof__(self) weakSelf = self;
    _imageUploadProgress = UploadImageAttachments(self.textView.attributedText, ^(NSString *plainText, NSError *error) {
        __typeof__(self) self = weakSelf;
        if (error) {
            [overlay dismiss:NO];
            
            [self enableEverything];
            
            if (([error.domain isEqualToString:NSCocoaErrorDomain] && error.code == NSUserCancelledError)) {
                [self focusInitialFirstResponder];
            } else {
                // In case we're covered up by subsequent view controllers (console message about "detached view controllers"), aim for our navigation controller.
                UIViewController *presenter = self.navigationController ?: self;
                [presenter presentViewController:[UIAlertController alertWithTitle:@"Image Upload Failed" error:error] animated:YES completion:nil];
            }
        } else {
            __weak __typeof__(self) weakSelf = self;
            [self submitComposition:plainText completionHandler:^(BOOL success) {
                __typeof__(self) self = weakSelf;
                [overlay dismiss:YES completion:^{
                    if (success) {
                        if (self.delegate) {
                            [self.delegate composeTextViewController:self didFinishWithSuccessfulSubmission:YES shouldKeepDraft:NO];
                        } else {
                            [self dismissViewControllerAnimated:YES completion:nil];
                        }
                    } else {
                        [self enableEverything];
                        [self focusInitialFirstResponder];
                    }
                }];
            }];
        }
    });
}

- (UIView *)viewToOverlay
{
    UINavigationController *navigationController = self.navigationController;
    if (navigationController) {
        return navigationController.topViewController.view;
    } else {
        return self.view;
    }
}

- (void)submitComposition:(NSString *)composition completionHandler:(void(^)(BOOL success))completionHandler
{
    [self doesNotRecognizeSelector:_cmd];
}

- (void)cancel
{
    if (self.delegate) {
        [self.delegate composeTextViewController:self didFinishWithSuccessfulSubmission:NO shouldKeepDraft:YES];
    } else {
        [self dismissViewControllerAnimated:YES completion:nil];
    }
}

- (void)setCustomView:(UIView<AwfulComposeCustomView> *)customView
{
    UIEdgeInsets textContainerInset = self.textView.textContainerInset;
    textContainerInset.top -= CGRectGetHeight(_customView.frame);
    [_customView removeFromSuperview];
    [_customView removeConstraint:_customViewWidthConstraint];
    _customViewWidthConstraint = nil;
    _customView = customView;
    if (!customView) {
        self.textView.textContainerInset = textContainerInset;
        return;
    }
    customView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.textView addSubview:customView];
    CGFloat height = CGRectGetHeight(customView.frame);
    textContainerInset.top += height;
    self.textView.textContainerInset = textContainerInset;
    NSDictionary *metrics = @{ @"height": @(height) };
    NSDictionary *views = NSDictionaryOfVariableBindings(customView);
    [self.textView addConstraints:
     [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[customView]"
                                             options:0
                                             metrics:metrics
                                               views:views]];
    [self.textView addConstraints:
     [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[customView(height)]"
                                             options:0
                                             metrics:metrics
                                               views:views]];
    _customViewWidthConstraint = [NSLayoutConstraint constraintWithItem:customView
                                                              attribute:NSLayoutAttributeWidth
                                                              relatedBy:NSLayoutRelationEqual
                                                                 toItem:nil
                                                              attribute:NSLayoutAttributeNotAnAttribute
                                                             multiplier:1
                                                               constant:CGRectGetWidth(self.view.bounds)];
    
    // If we're not in the view controller hierarchy, we might not yet have a width, in which case we'll add the constraint later.
    if (_customViewWidthConstraint.constant > 0) {
        [customView addConstraint:_customViewWidthConstraint];
    }
}

- (void)didTapCancel
{
    if (_imageUploadProgress) {
        [_imageUploadProgress cancel];
        _imageUploadProgress = nil;
        [MRProgressOverlayView dismissAllOverlaysForView:self.viewToOverlay animated:YES completion:^{
            [self enableEverything];
        }];
    } else {
        [self cancel];
    }
}

- (void)enableEverything
{
    [self updateSubmitButtonItem];
    self.textView.editable = YES;
    self.customView.enabled = YES;
}

- (void)focusInitialFirstResponder
{
    UIResponder *initialFirstResponder = self.customView.initialFirstResponder ?: self.textView;
    [initialFirstResponder becomeFirstResponder];
}

- (void)disableEverythingButTheCancelButton
{
    self.submitButtonItem.enabled = NO;
    [self.view endEditing:YES];
    self.textView.editable = NO;
    self.customView.enabled = NO;
}

- (void)beginObservingKeyboardNotifications
{
    NSNotificationCenter *noteCenter = [NSNotificationCenter defaultCenter];
    _keyboardWillShowObserver = [noteCenter addObserverForName:UIKeyboardWillShowNotification
                                                          object:nil
                                                           queue:[NSOperationQueue mainQueue]
                                                      usingBlock:^(NSNotification *note)
    {
        [self keyboardWillThingy:note];
    }];
    _keyboardWillHideObserver = [noteCenter addObserverForName:UIKeyboardWillHideNotification
                                                        object:nil
                                                         queue:[NSOperationQueue mainQueue]
                                                    usingBlock:^(NSNotification *note)
    {
        [self keyboardWillThingy:note];
    }];
}

- (void)endObservingKeyboardNotifications
{
    NSNotificationCenter *noteCenter = [NSNotificationCenter defaultCenter];
    if (_keyboardWillShowObserver) {
        [noteCenter removeObserver:_keyboardWillShowObserver];
        _keyboardWillShowObserver = nil;
    }
    if (_keyboardWillHideObserver) {
        [noteCenter removeObserver:_keyboardWillHideObserver];
        _keyboardWillHideObserver = nil;
    }
}

- (void)keyboardWillThingy:(NSNotification *)note
{
    NSTimeInterval duration = ((NSNumber *)note.userInfo[UIKeyboardAnimationDurationUserInfoKey]).doubleValue;
    UIViewAnimationCurve curve = ((NSNumber *)note.userInfo[UIKeyboardAnimationCurveUserInfoKey]).integerValue;
    UIViewAnimationOptions options = curve << 16;
    CGRect keyboardEndScreenFrame = ((NSValue *)note.userInfo[UIKeyboardFrameEndUserInfoKey]).CGRectValue;
    CGRect keyboardEndWindowFrame = [self.view.window convertRect:keyboardEndScreenFrame fromWindow:nil];
    CGRect keyboardEndTextViewFrame = [self.textView convertRect:keyboardEndWindowFrame fromView:nil];
    CGRect overlap = CGRectIntersection(keyboardEndTextViewFrame, self.textView.bounds);
    UIEdgeInsets contentInset = self.textView.contentInset;
    UIEdgeInsets indicatorInsets = self.textView.scrollIndicatorInsets;
    contentInset.bottom = CGRectGetHeight(overlap);
    indicatorInsets.bottom = CGRectGetHeight(overlap);
    [UIView animateWithDuration:duration delay:0 options:options animations:^{
        self.textView.contentInset = contentInset;
        self.textView.scrollIndicatorInsets = indicatorInsets;
    } completion:^(BOOL finished) {
        if ([note.name isEqualToString:UIKeyboardWillShowNotification]) {
            CGRect caretRect = [self.textView caretRectForPosition:self.textView.selectedTextRange.end];
            [self.textView scrollRectToVisible:caretRect animated:YES];
        }
    }];
}

- (void)viewWillLayoutSubviews
{
    [super viewWillLayoutSubviews];
    CGFloat width = CGRectGetWidth(self.view.bounds);
    if (_customViewWidthConstraint && width > 0) {
        _customViewWidthConstraint.constant = width;
        
        // If our view wasn't already in the hierarchy when the custom view was added, we didn't know how wide to make the custom view. Now we do.
        if (![self.customView.constraints containsObject:_customViewWidthConstraint]) {
            [self.customView addConstraint:_customViewWidthConstraint];
        }
        [self.view layoutSubviews];
    }
}

- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    [self endObservingKeyboardNotifications];
    [self endObservingTextChangeNotification];
}

#pragma mark - UIViewControllerRestoration

- (void)encodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super encodeRestorableStateWithCoder:coder];
    [coder encodeObject:self.textView.attributedText forKey:AttributedTextKey];
}

- (void)decodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super decodeRestorableStateWithCoder:coder];
    self.textView.attributedText = [coder decodeObjectForKey:AttributedTextKey];
    
    // -viewDidLoad gets called before -decodeRestorableStateWithCoder: and so the text color gets loaded from the saved attributed string. Reapply the theme after restoring state.
    [self themeDidChange];
    
    [self updateSubmitButtonItem];
}

static NSString * const AttributedTextKey = @"AwfulAttributedText";

@end

File: App/Composition/ComposeTextViewController.swift
//  ComposeTextViewController.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import MRProgress

class ComposeTextViewController: ViewController {
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        navigationItem.leftBarButtonItem = cancelButtonItem
        navigationItem.rightBarButtonItem = submitButtonItem
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    deinit {
        endObservingTextChangeNotification()
        endObservingKeyboardNotifications()
    }
    
    /// The composition text view. Set its text or attributedText property as appropriate.
    var textView: UITextView {
        return view as! UITextView
    }
    
    /// The button that submits the composition when tapped. Set its title property.
    fileprivate(set) lazy var submitButtonItem: UIBarButtonItem = {
        return UIBarButtonItem(title: "Submit", style: .plain, target: self, action: #selector(didTapSubmit))
    }()
    
    /// The button that cancels the composition when tapped. Set its title as appropriate.
    fileprivate(set) lazy var cancelButtonItem: UIBarButtonItem = {
        return UIBarButtonItem(title: "Cancel", style: .plain, target: self, action: #selector(didTapCancel))
    }()
    
    /// Tells a reasonable responder to become first responder.
    func focusInitialFirstResponder() {
        let responder = customView?.initialFirstResponder ?? textView
        responder.becomeFirstResponder()
    }
    
    /// Refreshes the submit button's enabled status.
    func updateSubmitButtonItem() {
        submitButtonItem.isEnabled = canSubmitComposition
    }
    
    /// Returns YES when the submission is valid and ready, otherwise NO. The default is to return YES when the textView is nonempty.
    var canSubmitComposition: Bool {
        return !textView.text.isEmpty
    }
    
    /**
        Called just before submission, offering a chance to confirm whether the submission should continue. The default implementation immediately allows submission.
     
        - parameter handler: A block to call after determining whether submission should continue, which takes as a parameter YES if submission should continue or NO otherwise.
     */
    func shouldSubmit(_ handler: @escaping (Bool) -> Void) {
        handler(true)
    }
    
    /// Returns a description of the process of submission, such as "Posting…".
    var submissionInProgressTitle: String {
        return "Posting…"
    }
    
    /**
        Do the actual work of submitting the composition. The default implementation crashes.
     
        - parameter composition: The composition with upload images having been replaced by appropriate textual equivalents.
        - parameter completion: A block to call once submission is complete; the block takes a single parameter, either YES on success or NO if submission failed.
     */
    func submit(_ composition: String, completion: @escaping (_ success: Bool) -> Void) {
        fatalError("\(type(of: self)) needs to implement \(#function)")
    }
    
    /// Called when the cancel button is tapped and no submission is in progress. The default implementation simply informs the delegate; overridden implementations can do so directly or call super as desired.
    func cancel() {
        if let delegate = delegate {
            delegate.composeTextViewController(self, didFinishWithSuccessfulSubmission: false, shouldKeepDraft: true)
        } else {
            dismiss(animated: true, completion: nil)
        }
    }
    
    /// A view that perches atop the textView, housing additional fields like a "Subject" field or a thread tag picker. Should be an instance of UIView, even though I'm not bothering to use the type system to enforce that right now because it's slightly too annoying.
    var customView: ComposeCustomView? {
        willSet {
            guard let old = customView as? UIView else { return }
            
            old.removeFromSuperview()
            
            customViewWidthConstraint?.isActive = false
            customViewWidthConstraint = nil
        }
        didSet {
            if let old = oldValue as? UIView , customView == nil {
                textView.textContainerInset.top -= old.bounds.height
            }
            
            guard let customView = customView as? UIView else { return }
            customView.translatesAutoresizingMaskIntoConstraints = false
            textView.addSubview(customView)
            
            textView.textContainerInset.top += customView.bounds.height
            
            textView.leadingAnchor.constraint(equalTo: customView.leadingAnchor).isActive = true
            textView.topAnchor.constraint(equalTo: customView.topAnchor).isActive = true
            customView.heightAnchor.constraint(equalToConstant: customView.bounds.height).isActive = true
            customViewWidthConstraint = customView.widthAnchor.constraint(equalToConstant: view.bounds.width)
            
            // If we're not in the view controller hierarchy, we might not yet have a width, in which case we'll add the constraint later.
            if let constraint = customViewWidthConstraint , constraint.constant > 0 {
                constraint.isActive = true
            }
        }
    }
    
    fileprivate var customViewWidthConstraint: NSLayoutConstraint?
    
    weak var delegate: ComposeTextViewControllerDelegate?
    
    fileprivate var menuTree: CompositionMenuTree!
    
    fileprivate func beginObservingTextChangeNotification() {
        guard textDidChangeObserver == nil else { return }
        textDidChangeObserver = NotificationCenter.default.addObserver(forName: NSNotification.Name.UITextViewTextDidChange, object: textView, queue: OperationQueue.main, using: { [weak self] (note: Notification) in
            self?.updateSubmitButtonItem()
        })
    }
    private func endObservingTextChangeNotification() {
        guard let token = textDidChangeObserver else { return }
        NotificationCenter.default.removeObserver(token)
        textDidChangeObserver = nil
    }
    private var textDidChangeObserver: NSObjectProtocol?
    
    fileprivate func beginObservingKeyboardNotifications() {
        guard keyboardWillShowObserver == nil else { return }
        
        keyboardWillShowObserver = NotificationCenter.default.addObserver(forName: NSNotification.Name.UIKeyboardWillShow, object: nil, queue: OperationQueue.main, using: { [weak self] (notification) in
            self?.keyboardWillThingy(notification)
        })
        
        keyboardWillHideObserver = NotificationCenter.default.addObserver(forName: NSNotification.Name.UIKeyboardWillHide, object: nil, queue: OperationQueue.main, using: { [weak self] (notification) in
            self?.keyboardWillThingy(notification)
        })
    }
    private func endObservingKeyboardNotifications() {
        if let token = keyboardWillShowObserver {
            NotificationCenter.default.removeObserver(token)
            keyboardWillShowObserver = nil
        }
        
        if let token = keyboardWillHideObserver {
            NotificationCenter.default.removeObserver(token)
            keyboardWillHideObserver = nil
        }
    }
    private var keyboardWillShowObserver: NSObjectProtocol?
    fileprivate var keyboardWillHideObserver: NSObjectProtocol?
    
    fileprivate func keyboardWillThingy(_ notification: Notification) {
        guard let
            duration = notification.userInfo?[UIKeyboardAnimationDurationUserInfoKey] as? TimeInterval,
            let curve = notification.userInfo?[UIKeyboardAnimationCurveUserInfoKey] as? UInt,
            let keyboardEndScreenFrame = (notification.userInfo?[UIKeyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue,
            let window = view.window
            else { return }
        let keyboardEndTextViewFrame = textView.convert(keyboardEndScreenFrame, to: window.screen.coordinateSpace)
        let overlap = keyboardEndTextViewFrame.intersection(textView.bounds)
        
        let options = UIViewAnimationOptions(rawValue: curve << 16)
        
        UIView.animate(withDuration: duration, delay: 0, options: options, animations: { 
            self.textView.contentInset.bottom = overlap.height
            self.textView.scrollIndicatorInsets.bottom = overlap.height
            }) { (finished) in
                guard let
                    endPosition = self.textView.selectedTextRange?.end
                    , notification.name == NSNotification.Name.UIKeyboardWillShow
                    else { return }
                let caretRect = self.textView.caretRect(for: endPosition)
                self.textView.scrollRectToVisible(caretRect, animated: true)
        }
    }
    
    fileprivate var imageUploadProgress: Progress?
    
    fileprivate func submit() {
        let overlay = MRProgressOverlayView.showOverlayAdded(to: viewToOverlay, title: submissionInProgressTitle, mode: .indeterminate, animated: true)
        overlay?.tintColor = theme["tintColor"]
        
        imageUploadProgress = uploadImages(attachedTo: textView.attributedText, completion: { [weak self] (plainText, error) in
            if let error = error {
                overlay?.dismiss(false)
                
                self?.enableEverything()
                
                if (error as NSError).domain == NSCocoaErrorDomain && (error as NSError).code == NSUserCancelledError {
                    self?.focusInitialFirstResponder()
                } else {
                    // In case we're covered up by subsequent view controllers (console message about "detached view controllers"), aim for our navigation controller.
                    let presenter = self?.navigationController ?? self
                    presenter?.present(UIAlertController.alertWithTitle("Image Upload Failed", error: error), animated: true, completion: nil)
                }
                
                return
            }
            
            guard let plainText = plainText else { fatalError("no error should mean yes plain text") }
            self?.submit(plainText, completion: { [weak self] (success) in
                if success {
                    if let delegate = self?.delegate {
                        delegate.composeTextViewController(self!, didFinishWithSuccessfulSubmission: true, shouldKeepDraft: false)
                    } else {
                        self?.dismiss(animated: true, completion: nil)
                    }
                } else {
                    self?.enableEverything()
                    
                    self?.focusInitialFirstResponder()
                    
                    return
                }
            })
        })
    }
    
    fileprivate var viewToOverlay: UIView {
        if let top = navigationController?.topViewController {
            return top.view
        } else {
            return view
        }
    }
    
    fileprivate func enableEverything() {
        updateSubmitButtonItem()
        
        textView.isEditable = true
        customView?.enabled = true
    }
    
    fileprivate func disableEverythingButTheCancelButton() {
        view.endEditing(true)
        
        submitButtonItem.isEnabled = false
        textView.isEditable = false
        customView?.enabled = false
    }
    
    @objc fileprivate func didTapSubmit() {
        disableEverythingButTheCancelButton()
        shouldSubmit({ [weak self] (ok: Bool) in
            if ok {
                self?.submit()
            } else {
                self?.imageUploadProgress?.cancel()
                self?.imageUploadProgress = nil
                
                self?.enableEverything()
                
                self?.focusInitialFirstResponder()
            }
        })
    }
    
    @objc fileprivate func didTapCancel() {
        if let progress = imageUploadProgress {
            progress.cancel()
            imageUploadProgress = nil
            
            MRProgressOverlayView.dismissAllOverlays(for: viewToOverlay, animated: true, completion: { 
                self.enableEverything()
            })
        } else {
            cancel()
        }
    }
    
    override func loadView() {
        let textView = ComposeTextView()
        textView.restorationIdentifier = "ComposeTextView"
        textView.font = UIFont.preferredFont(forTextStyle: UIFontTextStyle.body)
        textView.delegate = self
        view = textView
    }
    
    override func themeDidChange() {
        super.themeDidChange()
        
        textView.textColor = theme["listTextColor"]
        textView.keyboardAppearance = theme.keyboardAppearance
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        menuTree = CompositionMenuTree(textView: textView)
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        if let constraint = customViewWidthConstraint , view.bounds.width > 0 {
            constraint.constant = view.bounds.width
            
            // If our view wasn't already in the hierarchy when the custom view was added, we didn't know how wide to make the custom view. Now we do.
            constraint.isActive = true
            
            view.layoutSubviews()
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        focusInitialFirstResponder()
        
        updateSubmitButtonItem()
        
        beginObservingKeyboardNotifications()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        beginObservingTextChangeNotification()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        view.endEditing(true)
    }
    
    // MARK: State restoration
    
    override func encodeRestorableState(with coder: NSCoder) {
        super.encodeRestorableState(with: coder)
        
        coder.encode(textView.attributedText, forKey: Keys.AttributedText.rawValue)
    }
    
    override func decodeRestorableState(with coder: NSCoder) {
        super.decodeRestorableState(with: coder)
        
        textView.attributedText = coder.decodeObject(forKey: Keys.AttributedText.rawValue) as? NSAttributedString
        
        // -viewDidLoad gets called before -decodeRestorableStateWithCoder: and so the text color gets loaded from the saved attributed string. Reapply the theme after restoring state.
        themeDidChange()
        
        updateSubmitButtonItem()
    }
}

fileprivate enum Keys: String {
    case AttributedText
}

// For benefit of subclasses.
extension ComposeTextViewController: UITextViewDelegate {}

protocol ComposeCustomView {
    var enabled: Bool { get set }
    
    /// Returns a responder that should be the initial first responder when the AwfulComposeTextViewController first appears, instead of its textView. The default is nil, meaning the textView becomes first responder as usual.
    var initialFirstResponder: UIResponder? { get }
}

@objc protocol ComposeTextViewControllerDelegate: class {
    /**
        Sent to the delegate when composition is either submitted or cancelled.
     
        - parameter success: true if the submission was successful, otherwise false.
        - parameter keepDraft: true if the view controller should be kept around, otherwise false.
     */
    func composeTextViewController(_ composeController: ComposeTextViewController, didFinishWithSuccessfulSubmission success: Bool, shouldKeepDraft: Bool)
}
